File: connect4/connect4.py
"""Connect 4 — 7-column × 6-row vertical 4-in-a-row.

Same duck-typed interface as TicTacToe / ChessGame so it plugs into the
existing MCTS, model, and training loop without any of those needing changes.
Action = column index 0..6; piece falls to the lowest empty row.
Row 0 is the top of the board, row 5 is the bottom.
"""

import numpy as np


class Connect4:
    def __init__(self):
        self.row_count = 6
        self.column_count = 7
        self.action_size = self.column_count
        self.num_channels = 3  # current player, opponent, empty

    def get_initial_state(self):
        return np.zeros((self.row_count, self.column_count), dtype=np.int8)

    def update_state(self, state: np.ndarray, action: int, player: int) -> np.ndarray:
        """Drop a piece into column `action`. Mutates and returns state."""
        col = action
        for row in range(self.row_count - 1, -1, -1):
            if state[row, col] == 0:
                state[row, col] = player
                return state
        raise ValueError(f"Column {col} is full")

    def get_valid_moves(self, state: np.ndarray) -> np.ndarray:
        """A column is playable iff its top cell is empty."""
        return (state[0] == 0).astype(np.uint8)

    def check_win(self, state: np.ndarray, action: int) -> bool:
        """Did the piece most recently dropped into column `action` complete a 4-in-a-row?"""
        col = action
        # Topmost piece in the column = the one just dropped.
        row = None
        for r in range(self.row_count):
            if state[r, col] != 0:
                row = r
                break
        if row is None:
            return False
        player = state[row, col]
        # Four directions: horizontal, vertical, both diagonals.
        for dr, dc in [(0, 1), (1, 0), (1, 1), (1, -1)]:
            count = 1
            r, c = row + dr, col + dc
            while 0 <= r < self.row_count and 0 <= c < self.column_count and state[r, c] == player:
                count += 1
                r += dr
                c += dc
            r, c = row - dr, col - dc
            while 0 <= r < self.row_count and 0 <= c < self.column_count and state[r, c] == player:
                count += 1
                r -= dr
                c -= dc
            if count >= 4:
                return True
        return False

    def get_value_and_terminated(self, state: np.ndarray, action: int):
        if self.check_win(state, action):
            return 1, True
        if np.sum(self.get_valid_moves(state)) == 0:
            return 0, True
        return 0, False

    def get_opponent(self, player: int) -> int:
        return -player

    def state_hash(self, state: np.ndarray) -> int:
        return hash(state.tobytes())

    def encode_state(self, state: np.ndarray, player: int) -> np.ndarray:
        """3 channels from `player`'s perspective: own, opponent, empty."""
        encoded = np.zeros((3, self.row_count, self.column_count), dtype=np.float32)
        encoded[0] = (state == player).astype(np.float32)
        encoded[1] = (state == -player).astype(np.float32)
        encoded[2] = (state == 0).astype(np.float32)
        return encoded

File: web/src/connect4.ts
// Browser-side port of connect4/connect4.py.
//
// Same duck-typed interface as the Python class. Board is a flat
// Int8Array of length 42 in row-major order: index = r * 7 + c.
// Player is +1 or -1.

export class Connect4 {
  readonly rowCount = 6;
  readonly columnCount = 7;
  readonly actionSize = 7;
  readonly numChannels = 3;

  getInitialState(): Int8Array {
    return new Int8Array(this.rowCount * this.columnCount);
  }

  copy(state: Int8Array): Int8Array {
    return new Int8Array(state);
  }

  private at(state: Int8Array, r: number, c: number): number {
    return state[r * this.columnCount + c];
  }

  private setCell(state: Int8Array, r: number, c: number, v: number): void {
    state[r * this.columnCount + c] = v;
  }

  /** Drop a piece in column `action`. Mutates and returns state. */
  updateState(state: Int8Array, action: number, player: number): Int8Array {
    for (let r = this.rowCount - 1; r >= 0; r--) {
      if (this.at(state, r, action) === 0) {
        this.setCell(state, r, action, player);
        return state;
      }
    }
    throw new Error(`Column ${action} is full`);
  }

  getValidMoves(state: Int8Array): Uint8Array {
    const valid = new Uint8Array(this.actionSize);
    for (let c = 0; c < this.columnCount; c++) {
      valid[c] = this.at(state, 0, c) === 0 ? 1 : 0;
    }
    return valid;
  }

  /** Did the piece most recently dropped in `action` complete a 4-in-a-row? */
  checkWin(state: Int8Array, action: number): boolean {
    const col = action;
    let row = -1;
    for (let r = 0; r < this.rowCount; r++) {
      if (this.at(state, r, col) !== 0) { row = r; break; }
    }
    if (row === -1) return false;
    const player = this.at(state, row, col);
    const dirs: [number, number][] = [[0, 1], [1, 0], [1, 1], [1, -1]];
    for (const [dr, dc] of dirs) {
      let count = 1;
      let r = row + dr, c = col + dc;
      while (r >= 0 && r < this.rowCount && c >= 0 && c < this.columnCount && this.at(state, r, c) === player) {
        count++; r += dr; c += dc;
      }
      r = row - dr; c = col - dc;
      while (r >= 0 && r < this.rowCount && c >= 0 && c < this.columnCount && this.at(state, r, c) === player) {
        count++; r -= dr; c -= dc;
      }
      if (count >= 4) return true;
    }
    return false;
  }

  getValueAndTerminated(state: Int8Array, action: number): { value: number; terminated: boolean } {
    if (this.checkWin(state, action)) return { value: 1, terminated: true };
    const valid = this.getValidMoves(state);
    let any = 0;
    for (let i = 0; i < valid.length; i++) any |= valid[i];
    if (!any) return { value: 0, terminated: true };
    return { value: 0, terminated: false };
  }

  getOpponent(player: number): number {
    return -player;
  }

  /** Encode as 3 channels (own, opponent, empty), shape (3, 6, 7), C-order flat. */
  encodeState(state: Int8Array, player: number): Float32Array {
    const enc = new Float32Array(this.numChannels * this.rowCount * this.columnCount);
    const planeSize = this.rowCount * this.columnCount;
    for (let r = 0; r < this.rowCount; r++) {
      for (let c = 0; c < this.columnCount; c++) {
        const v = this.at(state, r, c);
        const idx = r * this.columnCount + c;
        if (v === player) enc[idx] = 1;
        else if (v === -player) enc[planeSize + idx] = 1;
        else enc[2 * planeSize + idx] = 1;
      }
    }
    return enc;
  }
}
